relay-config/src/config.rs

```diff
@@ -701,32 +704,127 @@ pub enum KafkaTopic {
+            KafkaTopic::Transactions => &self.transactions,
+            KafkaTopic::Outcomes => &self.outcomes,
+            KafkaTopic::Sessions => &self.sessions,
+            KafkaTopic::Metrics => &self.metrics,
+        }
+    }
+
+    /// Turn this mapping into a mapping of references
+    pub fn as_ref(&self) -> TopicMap<&T> {
+        TopicMap {
+            events: &self.events,
+            attachments: &self.attachments,
+            transactions: &self.transactions,
+            outcomes: &self.outcomes,
+            sessions: &self.sessions,
+            metrics: &self.metrics,
+        }
+    }
+
+    /// Apply `f` to convert values into different type.
+    pub fn map<U, E, F>(self, mut f: F) -> Result<TopicMap<U>, E>
+    where
+        F: FnMut(T) -> Result<U, E>,
+    {
+        Ok(TopicMap {
+            events: f(self.events)?,
+            attachments: f(self.attachments)?,
+            transactions: f(self.transactions)?,
+            outcomes: f(self.outcomes)?,
+            sessions: f(self.sessions)?,
+            metrics: f(self.metrics)?,
+        })
+    }
+}
+
+impl Default for TopicMap<TopicAssignment> {
     fn default() -> Self {
         Self {
-            events: "ingest-events".to_owned(),
-            attachments: "ingest-attachments".to_owned(),
-            transactions: "ingest-transactions".to_owned(),
-            outcomes: "outcomes".to_owned(),
-            sessions: "ingest-sessions".to_owned(),
-            metrics: "ingest-metrics".to_owned(),
+            events: "ingest-events".to_owned().into(),
+            attachments: "ingest-attachments".to_owned().into(),
+            transactions: "ingest-transactions".to_owned().into(),
+            outcomes: "outcomes".to_owned().into(),
+            sessions: "ingest-sessions".to_owned().into(),
+            metrics: "ingest-metrics".to_owned().into(),
+        }
+    }
+}
+
+/// Configuration for a "logical" topic/datasink that Relay should forward data into.
+///
+/// Can be either a string containing the kafka topic name to produce into (using the default
+/// `kafka_config`), or an object containing keys `topic_name` and `kafka_config_name` for using a
+/// custom kafka cluster.
+///
+/// See documentation for `secondary_kafka_configs` for more information.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+pub enum TopicAssignment {
+    /// String containing the kafka topic name. In this case the default kafka cluster configured
+    /// in `kafka_config` will be used.
+    Primary(String),
+    /// Object containing topic name and string identifier of one of the clusters configured in
+    /// `secondary_kafka_configs`. In this case that custom kafkaconfig will be used to produce
+    /// data to the given topic name.
+    Secondary {
+        /// The topic name to use.
+        topic_name: String,
+        /// An identifier referencing one of the kafka configurations in `secondary_kafka_configs`.
+        kafka_config_name: String,
```

Comment:
These should be named differently as per spec.

```suggestion
        name: String,
        /// An identifier referencing one of the kafka configurations in `secondary_kafka_configs`.
        broker: String,
```

Reply:
I wanted to change this because I found our use of "kafka config" vs "broker" inconsistent. The example where we write `secondary_kafka_config: some_broker: ...` only looks good because the *user* named their kafka config "broker", but it can be any string

Reply:
Makes sense, I prefer @untitaker's proposal (I vote for changing the spec).

Reply:
Sounds good, thanks. I'm 👍  with this.

In an attempt to keep it short, a last suggestion on the proposal: `name` and `config`. Given that this is in a key called `topics`, I think `name` is unambiguous.

Reply:
applied